utils/s3/client.cc

```diff
@@ -1092,6 +1093,112 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
+                                s3l.trace("Fiber for object '{}' will try to download within range {}-{}", _object_name, _range.off, _range.len);
+                                auto buff_size = buf.size();
+                                _range.off += buff_size;
+                                _range.len -= buff_size;
+                                _buffers_size += buff_size;
+                                s3l.trace("Fiber for object '{}' pushes {} bytes buffer", _object_name, buff_size);
+                                _buffers.emplace_back(std::move(buf), co_await _client->claim_memory(buff_size), s3_clock::now() - start);
+                                _buffer_cv.broadcast();
+                                if (buff_size == 0) {
+                                    s3l.trace("Fiber for object '{}' signals EOS", _object_name);
+                                    co_return consumption_result(stop_consuming(temporary_buffer<char>()));
+                                }
+                                co_await coroutine::maybe_yield();
+                                recap_start = s3_clock::now();
+                                co_return consumption_result<char>(continue_consuming());
+                            }
+                            co_return consumption_result(stop_consuming(temporary_buffer<char>()));
+                        });
+                    },
+                    {},
+                    _as);
+                co_await _buffer_cv.when();
+            } catch (...) {
+                s3l.trace("Fiber for object '{}' failed: {}, exiting", _object_name, std::current_exception());
+                _buffer_cv.broken(std::current_exception());
+                co_return;
+            }
+        }
+    }
+
+public:
+    chunked_download_source(shared_ptr<client> cln, sstring object_name, std::optional<range> range, seastar::abort_source* as)
+        : _client(std::move(cln)), _object_name(std::move(object_name)), _as(as), _range(range.value_or(s3::range{0, std::numeric_limits<uint64_t>::max()})) {
+        _filling_fiber = make_filling_fiber();
+        _buffer_cv.signal();
+        s3l.trace("Constructing chunked_download_source for object '{}'", _object_name);
+    }
+
+    future<temporary_buffer<char>> get() override {
+        while (true) {
+            if (!_buffers.empty()) {
+                auto claimed_buff = std::move(_buffers.front());
+                _buffers.pop_front();
+                _buffers_size -= claimed_buff._buffer.size();
+                _buffer_cv.broadcast();
+                s3l.trace("get() for object '{}' popped buffer of {} bytes", _object_name, claimed_buff._buffer.size());
+                // Report the time it took to consume the buffer here and not in the fiber, otherwise our read amplification checks will go haywire
+                _client->find_or_create_client().read_stats.update(claimed_buff._buffer.size(), claimed_buff._consumption_time);
```

Comment:
We have to lie about our stats and not report it in the downloading fiber, otherwise, because of the buffering we will get huge read amplification in our stats

Reply:
But that's not lie, it's _exactly_ what's going on with S3 client and stats should reflect that

Reply:
That's not a lie, it is a terrible lie :)
This is _not_ whats going on with s3 client, the client reads from S3, but these bytes may not be consumed, so you are missing your stats about number of reads done against S3 server and bytes transferred. If we want to `get` only two bytes and the fiber already downloaded it's `_max_buffers_size` bytes from S3 and then you report only 2 bytes being read from S3 then it changes the whole definition of truth

Reply:
>  If we want to get only two bytes and the fiber already downloaded it's _max_buffers_size bytes from S3 and then you report only 2 bytes being read from S3 then it changes the whole definition of truth

Why "then you report only 2 bytes"? Here you report full buffer size, not 2 bytes. And my point is that we should report full buffer size as well, but not here -- in the GET reply hander

Reply:
A. -- yes this is what S3 metrics should show
B. -- no, it's not what we want, at least not on S3 client. Compressed sstables consumed from disk do count the compressed bytes, but not uncompressed, so shy should S3 client care?
C. -- this value is zero, any data read from the socket gets into the _buffers (unless exception occurs)

Reply:
I guess the confusion here is about data size vs request latency. Before this patch (in fact -- before the download_source itself) the completed request carried along _all_ its data. So we updated both -- request latency and request data size -- in one place. Now these two bits are separate. Request completes and data arrives later.

Is that the case, @kreuzerkrieg ?